 ## Prepared By MIRZA RAZIN ALAM 101329992

**Problem 2.1**

a) The task required is supervised because this a classification problem. We know about the types of expected output.

b) The task required here is unsupervised because this requires a data filtering and no specific outcome is predicted or classified.

c) This is a supervised learning task because it requires classification.

d) This is an unsupervised learning task because we are looking for patterns within the input data.

e) Since we trying to predict an output, the task in hand is related to supervised learning.

f) This is supervised learning as we are trying to make a prediction.

g) This is a supervised learning task as it requires classification.

h) This is an unsupervised learning task as it requires analysis of the input data and there is no outcome variable. This requires association or affinity analysis.

**Problem 2.5**

Overfitting occurs when there is excess levels of data being used in the training model. Although the error is equal to zero, 
overfitting can cause the model to make bad predictions and fails to generalize.

**Problem 2.8**

In [3]:
# import pandas and numpy libraries
import numpy as np
import pandas as pd
# create a data frame
df = pd.DataFrame({
'Age': [25,56,65, 32,41,49],
'Income': [49000, 156000, 99000, 192000, 39000, 57000]
})
print(df)

   Age  Income
0   25   49000
1   56  156000
2   65   99000
3   32  192000
4   41   39000
5   49   57000


In [6]:
# import preprocessing
# Normalizing a data frame in sci-kit learn
from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler
# scikit-learn:
scaler = StandardScaler()
norm_df2 = pd.DataFrame(scaler.fit_transform(df),
index=df.index, columns=df.columns)
# the result of the transformation is a numpy array, we convert it into a dataframeprint(norm_df2)
print(norm_df2)

        Age    Income
0 -1.438597 -0.865431
1  0.829022  0.999021
2  1.487363  0.005808
3 -0.926554  1.626314
4 -0.268213 -1.039679
5  0.316979 -0.726033


**Problem 2.10**

We will use model B for final deployment because model A seems to be overfitted and therefore had better accuracy with the training data.

**Problem 2.11 A**

Here the categorical variables 'Fuel_Type' and 'Color' is denoted by string values such as 'Diesel', 'Blue' etc., we can sometimes code the categories numerically (1, 2, 3, …) and treat the variable as if it were a continuous variable. 

In [7]:
#Loading the csv file and creating the initial dataframe df
df= pd.read_csv('ToyotaCorolla(4).csv')
df

Id                                              Model  Price  \
0        1      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500   
1        2      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750   
2        3      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950   
3        4      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950   
4        5        TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750   
...    ...                                                ...    ...   
1431  1438         TOYOTA Corolla 1.3 16V HATCHB G6 2/3-Doors   7500   
1432  1439  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...  10845   
1433  1440  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   8500   
1434  1441  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   7250   
1435  1442        TOYOTA Corolla 1.6 LB LINEA TERRA 4/5-Doors   6950   

      Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  ...  \
0            23         10      2002  46986    Diesel   90          1  ...   
1            23         10      2002  72937    Diesel   90          1  ...   
2            24          9      2002  41711    Diesel   90          1  ...   
3            26          7      2002  48000    Diesel   90          0  ...   
4            30          3      2002  38500    Diesel   90          0  ...   
...         ...        ...       ...    ...       ...  ...        ...  ...   
1431         69         12      1998  20544    Petrol   86          1  ...   
1432         72          9      1998  19000    Petrol   86          0  ...   
1433         71         10      1998  17016    Petrol   86          0  ...   
1434         70         11      1998  16916    Petrol   86          1  ...   
1435         76          5      1998      1    Petrol  110          0  ...   

     Powered_Windows  Power_Steering  Radio  Mistlamps  Sport_Model  \
0                  1               1      0          0            0   
1                  0               1      0          0            0   
2                  0               1      0          0            0   
3                  0               1      0          0            0   
4                  1               1      0          1            0   
...              ...             ...    ...        ...          ...   
1431               1               1      0          1            1   
1432               0               1      0          0            1   
1433               0               1      0          0            0   
1434               0               0      0          0            0   
1435               0               1      0          0            0   

      Backseat_Divider  Metallic_Rim  Radio_cassette  Parking_Assistant  \
0                    1             0               0                  0   
1                    1             0               0                  0   
2                    1             0               0                  0   
3                    1             0               0                  0   
4                    1             0               0                  0   
...                ...           ...             ...                ...   
1431                 1             0               0                  0   
1432                 1             0               0                  0   
1433                 1             0               0                  0   
1434                 1             0               0                  0   
1435                 0             0               0                  0   

      Tow_Bar  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1431        0  
1432        0  
1433        0  
1434        0  
1435        0  

[1436 rows x 39 columns]

In [16]:
# create dummy variables
df_dummy = pd.get_dummies(df, columns=['Fuel_Type','Color'],prefix_sep='_', drop_first=True)

In [21]:
#Summary of df_dummy
df_dummy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 48 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Id                 1436 non-null   int64 
 1   Model              1436 non-null   object
 2   Price              1436 non-null   int64 
 3   Age_08_04          1436 non-null   int64 
 4   Mfg_Month          1436 non-null   int64 
 5   Mfg_Year           1436 non-null   int64 
 6   KM                 1436 non-null   int64 
 7   HP                 1436 non-null   int64 
 8   Met_Color          1436 non-null   int64 
 9   Automatic          1436 non-null   int64 
 10  CC                 1436 non-null   int64 
 11  Doors              1436 non-null   int64 
 12  Cylinders          1436 non-null   int64 
 13  Gears              1436 non-null   int64 
 14  Quarterly_Tax      1436 non-null   int64 
 15  Weight             1436 non-null   int64 
 16  Mfr_Guarantee      1436 non-null   int64 


In [19]:
#Checking all the variables in the new dataframe df_dummy
print(df_dummy.loc[:, 'Fuel_Type_Diesel':'Color_Yellow'].head(5))

   Fuel_Type_Diesel  Fuel_Type_Petrol  Color_Black  Color_Blue  Color_Green  \
0                 1                 0            0           1            0   
1                 1                 0            0           0            0   
2                 1                 0            0           1            0   
3                 1                 0            1           0            0   
4                 1                 0            1           0            0   

   Color_Grey  Color_Red  Color_Silver  Color_Violet  Color_White  \
0           0          0             0             0            0   
1           0          0             1             0            0   
2           0          0             0             0            0   
3           0          0             0             0            0   
4           0          0             0             0            0   

   Color_Yellow  
0             0  
1             0  
2             0  
3             0  
4             0  


**Problem 2.11 B**


In [34]:
from sklearn.model_selection import train_test_split

# random_state is set to a defined value to get the same partitions when re-running
# the code

# training 
trainData, temp = train_test_split(df_dummy, test_size=0.5, random_state=1)
validData, testData = train_test_split(temp, test_size=0.4, random_state=1)
print('Training : ', trainData.shape)
print('Validation : ', validData.shape)
print('Test : ', testData.shape)


Training :  (718, 48)
Validation :  (430, 48)
Test :  (288, 48)


718 observations of data will be used for training the model, which contains the data used to build the various models we are examining. 430 observations of data will be used to assess the predictive performance of each model so that you can compare models and choose the best one. Finally, 288 observations of data is used to assess the performance of the chosen model with new data.